core/src/main/java/org/eclipse/angus/mail/util/MailSSLSocketFactory.java

```diff
@@ -334,51 +336,120 @@ public synchronized Socket createSocket(String s, int i)
-        private X509TrustManager adapteeTrustManager = null;
+        private final X509ExtendedTrustManager adapteeTrustManager;
 
-        /**
-         * Initializes a new TrustManager instance.
-         */
         private MailTrustManager() throws GeneralSecurityException {
             TrustManagerFactory tmf = TrustManagerFactory.getInstance("X509");
             tmf.init((KeyStore) null);
-            adapteeTrustManager = (X509TrustManager) tmf.getTrustManagers()[0];
+
+            TrustManager tm = tmf.getTrustManagers()[0];
+
+            if (tm instanceof X509ExtendedTrustManager) {
+                adapteeTrustManager = (X509ExtendedTrustManager) tm;
+            } else {
+                // Should not happen, but wrap X509TrustManager if needed
+                adapteeTrustManager = new X509ExtendedTrustManager() {
+                    private final X509TrustManager base = (X509TrustManager) tm;
+
+                    @Override
+                    public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket)
+                            throws CertificateException {
+                        base.checkClientTrusted(chain, authType);
+                    }
+
+                    @Override
+                    public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+                            throws CertificateException {
+                        base.checkClientTrusted(chain, authType);
+                    }
+
+                    @Override
+                    public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket)
+                            throws CertificateException {
+                        base.checkServerTrusted(chain, authType);
+                    }
+
+                    @Override
+                    public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+                            throws CertificateException {
+                        base.checkServerTrusted(chain, authType);
+                    }
+
+                    @Override
+                    public void checkClientTrusted(X509Certificate[] chain, String authType)
+                            throws CertificateException {
+                        base.checkClientTrusted(chain, authType);
+                    }
+
+                    @Override
+                    public void checkServerTrusted(X509Certificate[] chain, String authType)
+                            throws CertificateException {
+                        base.checkServerTrusted(chain, authType);
+                    }
+
+                    @Override
+                    public X509Certificate[] getAcceptedIssuers() {
+                        return base.getAcceptedIssuers();
+                    }
+                };
+            }
+        }
+
+        private boolean shouldDelegate() {
+            return !(isTrustAllHosts() || getTrustedHosts() != null);
```

Comment:
Double check me on this, but i think for X509ExtendedTrustManager methods this is not correct when EIA is enabled and we are trusting a list of hosts.  In that case we should short circuit if matching, otherwise delegate.

Essentially we should call MailSSLSocketFactory::isServerTrusted to determine if we delegate. Or rather update this private method to take a host name.

Reply:
We might have to remove MailSSLSocketFactory::isServerTrusted call from SocketFetcher and just have the call happen in the trustmanager. The reason I think we have to do that is because we don't reflective invoke EIA from SocketFetcher.